[TAN-5473] Avoid bulk invites import timeouts by using background jobs

jinjagit opened the pull request:
- uses 2 new background jobs: `CountNewSeatsJob` & `BulkCreateJob`
- creates new `invites_imports` db table, used to hold data to poll for result of completed jobs, and related model, endpoint, etc.
- fixes bug where spinner on submit button only showed while counting seats but not while creating invites.
- functionally, no change in UX (except spinner bug fixed) - still reports errors per xlsx row, etc.
- adds tests of background jobs && of `useInvitesImport`

# Changelog
## Fixed
- [TAN-5473] Avoid bulk invites import timeouts (by using background jobs)

jinjagit committed:
[TAN-5473] basic invites_import table + model

jinjagit committed:
[TAN-5473] Remove restrict command from structure.sql

jinjagit committed:
[TAN-5473] Remove unrestrict command from structure.sql

jinjagit committed:
Merge branch 'master' into TAN-5473-use-background-jobs-for-bulk-invites

jinjagit committed:
[TAN-5473] Basic implementation of CountNewSeatsJob

jinjagit committed:
[TAN-5473] count_new_seats_xlsx method in job

jinjagit committed:
[TAN-5473] Also use CountNewSeatsJob for count_new_seats endpoint

jinjagit committed:
[TAN-5473] Add show endpoint for invites_imports

jinjagit committed:
[TAN-5473] Linting

jinjagit committed:
[TAN-5473] Linting (again)

jinjagit committed:
[TAN-5473] Set completed_at in CountNewSeatsJob

jinjagit committed:
[TAN-5473] Some basic model tests

jinjagit committed:
[TAN-5473] Linting

jinjagit committed:
[TAN-5473] Align data in result

jinjagit committed:
[TAN-5473] Fix count_seats acceptance tests

jinjagit committed:
[TAN-5473] Add tests that CountNewseatsJob is enqueued

jamesspeake committed:
[TAN-5473] adding polling for bulk invite jobs

jinjagit committed:
[TAN-5473] Linting

jinjagit committed:
[TAN-5473] More linting

jinjagit committed:
Merge pull request #12130 from CitizenLabDotCo/TAN-5473-JS-use-background-jobs-for-bulk-invites

TAN-5473 - WIP add polling for bulk invite jobs

jinjagit committed:
[TAN-5473] Remove unrelated changes to structure.sql

jinjagit committed:
[TAN-5473] Add enabled option to useInviteImports to avoid redundant requests

jinjagit committed:
[TAN-5473] Also use new polling for count seats when emails given manually (not in xlsx)

jinjagit committed:
[TAN-5473] FE linting

jinjagit committed:
[TAN-5473] FE linting + fix infinite request loop

jinjagit committed:
[TAN-5473] Remove weird pack-lock.json file

jinjagit committed:
[TAN-5473] Prevent stale data causing confirm seats modal to reappear when remount component

jinjagit committed:
[TAN-5473] Linting - add useCallback to onSubmit & reorder functions accordingly

jinjagit committed:
[TAN-5473] Linting - add useCallback to 3 more functions & reorder functions accordingly

jinjagit committed:
Reset CI trigger after force push

jinjagit committed:
[TAN-5473] New BulkCreateJob, used in InvitesController + very basic tests

jinjagit committed:
[TAN-5473] Also use polling approach for create invites step

jinjagit committed:
[TAN-5473] Linting - remove  unnecessary dependencies

jinjagit committed:
[TAN-5473] Use job_type attribute in conditional to prevent double messages

jinjagit committed:
[TAN-5473] Read errors from responses correctly

jinjagit committed:
[TAN-5473] Rename some vars

jinjagit committed:
[TAN-5473] Add some comments

jinjagit committed:
Merge branch 'master' into TAN-5473-use-background-jobs-for-bulk-invites

jinjagit committed:
[TAN-5473] Basic tests of CountNewSeatsJob

jinjagit committed:
[TAN-5473] Linting

jinjagit committed:
[TAN-5473] More linting

jinjagit committed:
[TAN-5473] Remove redundant vars

jinjagit committed:
[TAN-5473] Happy path tests of BulkCreateJob for manual emails

jinjagit committed:
[TAN-5473] Add tests that CountNewSeatsJob does not create users or invites

jinjagit committed:
[TAN-5473] Sad path tests of BulkCreateJob for manual emails

jinjagit committed:
[TAN-5473] Improve happy path tests of BulkCreateJob for manual emails

jinjagit committed:
[TAN-5473] Use shared examples in count_new_seats_job_spec

jinjagit committed:
[TAN-5473] Simplify shared example

jinjagit committed:
[TAN-5473] bulk_creat_job_spec

jinjagit committed:
[TAN-5473] Use shared examples in bulk_create_job_spec

jinjagit committed:
[TAN-5473] Linting

jinjagit committed:
[TAN-5473] Raise MAX_INVITES to 10_000

jinjagit committed:
[TAN-5473] Use correct job type in FE test

jinjagit committed:
[TAN-5473] Improve invites_spec

jinjagit committed:
[TAN-5473] Basic invites_imports_spec

jinjagit committed:
[TAN-5473] Fix spinner not showing while prcessing invites

jinjagit commented on front/app/containers/Admin/invitations/invite/index.tsx:
```diff
@@ -470,7 +648,7 @@ const Invitations = () => {
               />
 
-              {processing && (
+              {(processing || createInviteImportId !== null) && (
```
Fixes existing bug, where spinner only showed while counting seats but not while creating invites.

jinjagit committed:
[TAN-5473] Bugfix: ensure correct seats displayed after invites change seat numbers

jinjagit committed:
[TAN-5473] FE linting

jinjagit committed:
[TAN-5473] Improve comments

jinjagit committed:
[TAN-5473] Remove the || 0 'fudge'

jinjagit committed:
[TAN-5473] Update job_type in test

jinjagit committed:
[TAN-5473] Change name of interface

jinjagit committed:
[TAN-5473] Fix job_type in test

jinjagit committed:
[TAN-5473] Rnename InviteImports to InvitesImport

jinjagit committed:
[TAN-5473] Add test of changes to seats in app_configuration

jamesspeake commented on back/app/controllers/web_api/v1/invites_controller.rb:
```diff
@@ -67,51 +67,86 @@ def index_xlsx
+      bulk_create_params,
+      import.id,
+      xlsx_import: false
+    )
```
Slightly wondering whether any of the methods that return serialized `InvitesImports` should actually sit in the invites_imports_controller?

jinjagit replied:
Good point. We discussed and agreed these endpoints should now, in fact, live in the `InvitesImportsController`.

jamesspeake commented on front/app/containers/Admin/invitations/invite/index.tsx:
```diff
@@ -460,7 +647,7 @@ const Invitations = () => {
             <Box display="flex" alignItems="center" paddingTop="30px">
               <SubmitWrapper
-                loading={processing}
+                loading={processing || createInvitesImportId !== null}
```
Wondering why you need to do this. Should we not be setting processing to true when `createInvitesImportId !== null`?

jinjagit replied:
I think this prevents a second message appearing alongside the first. I found this was a difficult issue to fix and this was the only way I found to not have the count seats step show a success message, while the create invites step would then create a new instance of the processing message alongside it.

This conditional fixed that issue.

jamesspeake commented on front/app/api/invites/useInvitesImport.ts:
```diff
@@ -0,0 +1,48 @@
+    queryKey: invitesKeys.lists(),
+    queryFn: () => fetchInvitesImport(queryParams),
+    refetchInterval: pollingEnabled ? 5000 : false,
+    enabled: isEnabled,
```
Feels like this hook could be simplified - looking at how it is called, so long as there is an importId, the hook should both be enabled and polling. And that param gets reset to nil in the calling component I think. So maybe you could get rid of the `pollingEnabled` and `enabled` params? And just use `!!queryParams.importId`? Could remove some complexity.

jamesspeake commented on front/app/containers/Admin/invitations/invite/index.tsx:
```diff
@@ -88,17 +97,355 @@ const Invitations = () => {
+  // State variables for bulk invite creation invites_import polling
+  const [createInvitesImportId, setCreateInvitesImportId] = useState<
+    string | null
+  >(null);
```
Wondering if you thought about just using the one `useInvitesImport` but updating the importId. Given they are used sequentially then it feels like you could? Might make it simpler.

jinjagit replied:
I tried that approach, but in the end found I needed to use `if (jobType === 'bulk_create' || jobType === 'bulk_create_xlsx')` in a couple of places.

jamesspeake commented on front/app/containers/Admin/invitations/invite/index.tsx:
```diff
@@ -88,17 +97,355 @@ const Invitations = () => {
+        try {
+          setProcessing(true);
+          const createJob = await bulkInviteEmails(savedEmailOptions);
+          setCreateInvitesImportId((createJob as any).data.id);
```
Should we not have a correct type here?

jinjagit replied:
Good catch. Will see if I can fix this.

jinjagit replied:
Fixed, and also noticed I was still using `IInvites` in the mutation hooks, `useBulkInviteEmails` && `useBulkInviteXLSX`. Changed it to `IInvitesImport` and then was able to remove the `as any`.

I wonder if this change makes it easier to simplify my changes in `invite/index.tsx` 🤞 
I'm not actually sure how it even worked without these changes!

jamesspeake reviewed (changes requested):
Generally looks good, just think maybe there is some simplification that can be done in the FE hooks. Might be worth just talking it through with you - it's also a little difficult to see exactly what changed as I think you've rearranged some of the code

jinjagit committed:
Merge branch 'master' into TAN-5473-use-background-jobs-for-bulk-invites

jinjagit committed:
[TAN-5473} Fix specs after merging master with bugfix for billed_admins

jinjagit committed:
[TAN-5473] Move count new seats endpoints to new controller

jinjagit committed:
[TAN-5473] Move bulk create endpoints to new controller

jinjagit committed:
[TAN-5473] Add accpetance tests of unauthorized requests

jinjagit committed:
[TAN-5473] Use new endpoints in FE

jinjagit committed:
[TAN-5473] Use new endpoints in FE tests

jinjagit committed:
[TAN-5473] Do not change MAX_INVITES = 1_000

jinjagit committed:
[TAN-5473] Use IInvitesImport, not IInvites, in mutation hooks

jinjagit committed:
Merge branch 'master' into TAN-5473-use-background-jobs-for-bulk-invites

jinjagit commented:
@jamesspeake I made the changes we discussed...

1. Moved `count_new_seats`, `count_new_seats`, `bulk_create`, and `bulk_create_xlsx` actions to the new `InvitesImportsController` + changed routes, paths, specs, etc accordingly.
2. Fixed the uses of `any` type, which also revealed an error by me (also fixed). See [my comment](https://github.com/CitizenLabDotCo/citizenlab/pull/12116/#discussion_r2379151179) for details.
3. Reverted to the original `MAX_INVIES = 1_000`, as 10k didn't seem to work locally (never completed count seats step, so I wonder if there is a row limit on xlsx we can process). 1k works fine, and is pretty snappy, so I figure just having that work robustly is a pretty good improvement.

jamesspeake committed:
[TAN-5473] Reduced code and fixed types for invites import

jamesspeake committed:
[TAN-5473] Reduced code and fixed types for invites import

jamesspeake committed:
[TAN-5473] Fixed deadcode

jamesspeake reviewed (approved):
Changes look good to me. Have shared a FE PR with the changes I was suggesting

jinjagit committed:
Merge pull request #12206 from CitizenLabDotCo/TAN-5473-JS-use-background-jobs-for-bulk-invites

TAN-5473 - Simplify and type invite imports
